### DISCUSSION TITLE
HASPmota Change Page using GPIO0 or 14 button

### DISCUSSION THREAD

**Discussion by richardsg307** [arendst/Tasmota#20316] [Q&A] [answered]
Running Tasmota, Berry and HASPmota on Lilygo T-Display-S3 ESP32 with a Non-Touch screen.
I'd like to be able to have several HASPmota pages and scroll through them with the GPIO0 (or 14) button (ie the buttons already on the board).
I'm guessing a Berry function could be triggered by the button press and the function then issues a 'Next' page command to HASPmota.
I don't know how to do either, and cannot find examples.

Edit: The penny dropped regarding triggering a Berry function. This code worked from the console
```
def rule_gpio0(value, trigger)
  print("GPIO0 button pressed - ", value)
end
tasmota.add_rule("Button1#Action", rule_gpio0)
```
and prints (with a single press)
```
GPIO0 button pressed -  SINGLE
```
So that just leaves the command to HASPmota to go to the next (or Previous) page to find.

**Comment by sfromis** [comment, marked as answer]
In Berry, you can select another page by calling the show method on the page object, like `p2.show()`

If you are maintaining the page number in the variable `page`, you could code like this in Berry:
`global.member(f'p{page}').show()` to reference the page object, and then call the `show()` method.


Note that the "primary" way of reacting to Buttons would be to trigger on `Button1#state`, but that also depends on what other button configs you've made, and you may still need more than the mentioned trigger if you want to avoid the Button input triggering default actions in Tasmota.

**Reply by richardsg307** [reply]
Ah, OK. I'll try that and use a variable so that Berry knows which page page I'm on so it can alternate back and forth.

Are you saying the `tasmota.add_rule("Button1#state", rule_gpio0)` would be better programming practice?. Using `Action `I get `SINGLE`, `DOUBLE`, `TRIPLE`, `QUAD`, `PENTA`, `HOLD `and `CLEAR` as the printed value ,and it does what I'm after.

**Reply by s-hadinger** [reply]
I realize there is no programmatic way to do next/prev page. I will add this

**Reply by sfromis** [reply]
I'm not talking about better/worse, just suggesting that what I mentioned is in principle the simplest, being available without needing `SetOption73`. However, with the caveat about default actions still applying unless you do something to defeat that. But no problem doing this with `SetOption73`, actually ;-) 

Even with `SetOption73`, using `tasmota.add_rule('Button1#state', ...)` will be very slightly faster, but only a few milliseconds, not really noticeable.

**Reply by richardsg307** [reply]
I've put a blank p2 in my pages .jsonl file
```
{"page":2,"comment":"---------- Page 2 Can be updated dynamically for experimentation ----------"}
```
 and added this in my autoexec.be
```
var cp = 1 # Current Page
def rule_gpio0(value, trigger)
  print("GPIO0 button pressed - ", value, trigger)
  if cp == 1 
    p2.show()
    cp = 2
  else
    p1.show()
    cp = 1
  end
end
tasmota.add_rule("Button1#Action", rule_gpio0)
``` 
It's not loading properly and I have the error 
```
BRY: Exception> 'syntax_error' - input:1: 'p1' undeclared (first use in this function)
```
Presumably it's trying to parse this before HASPmota has done its bit. What's the best way round this?

**Reply by sfromis** [reply]
You could use the multi-page technique, I suggested, does not depend on vars already available.

With the flipping technique you posted now, you could use `global.p2` instead of just `p2`. When making the global reference explicitly, instead of letting Berry infer this, it will no longer check if the name is already defined, as it does not have to infer whether it is a global or not.

Another approach could be to defer the rule creation till later, but that just makes it more complex.

BTW. Instead of your 7 lines of code, you could also:
```
(cp == 1 ? global.p2 : global.p1).show()
cp = 3 - cp
```
Of course, the technique with `global.member` also works.

**Reply by richardsg307** [reply]
I saw you slip that edit in - just as I was thinking same ;)
Yep, global works now as does the button changing the pages.
Final code for the page changing on the GPIO0 button press is
```
var cp = 1 # Current Page
def rule_gpio0(value, trigger)
  print("GPIO0 button pressed - ", value, trigger)
  if cp == 1 
    global.p2.show()
    cp = 2
  else
    global.p1.show()
    cp = 1
  end
end
tasmota.add_rule("Button1#Action", rule_gpio0)
```
BTW I tried your 2 lines of code (which look OK to me) but get the error
```
17:27:12.582 MQT: stat/boilerdisplay/RESULT = {"Button1":{"Action":"SINGLE"}}
17:27:12.686 GPIO0 button pressed -  SINGLE Button1#Action
17:27:12.788 BRY: Exception> 'type_error' - 'nil' value is not callable
17:27:12.889 stack traceback:
17:27:12.989 	<unknown source>: in function `rule_gpio0`
17:27:13.090 	<unknown source>: in function `try_rule`
17:27:13.191 	<unknown source>: in function `exec_rules`
17:27:13.292 	<unknown source>: in function `event`
```
whenever I press the button. Also, I do have Option73 On as I have other stuff going on communicating with Home Assistant over MQTT.

Big thanks to you both for all the effort you put into this and helping the community.

**Reply by sfromis** [reply]
Actually, I had verified the code suggestion as actually working, when having both pages already available.

That error suggests that you had the code execute without p1/p2 being available, like HASPmota not started yet, which is as bad with either of the coding techniques, but does no more harm than produce the "less nice" error message. The thing is that referring to a global not created yet will give you a `nil` instead of a direct error message.

**Reply by richardsg307** [reply]
Strange. Your code works from the berry console, just not in the function called from the button press. Thanks again for your help.

**Reply by sfromis** [reply]
The code suggestion always works, also in that context, but it needs to placed in a way where the syntax is valid for Berry.

I assume that you are placing it right after this:
`print("GPIO0 button pressed - ", value, trigger)`

As Berry has free-form syntax, without line breaks mattering, that means that the next
`(cp == 1 ? global.p2 : global.p1)` 
is seen as argument to the value returned by the `print()`. As the return value there is always empty (`nil`), it will then try call the `nil` as a function with that argument, and then it would call `show()` from the return from that, but it does obviously not get to that point.

In the syntax of Berry, the free format means that you can usually omit the `;` which works as a statement delimiter. It is rarely *needed*, except to disambiguate the syntax. Meaning that if you add a `;` after the print, the syntax will be clear, also to the Berry parser. Other ways of having a syntax where a ( would not bee seen as a continuation of the statement could also work, but the `;` is the simplest way.

Another "workaround" is that it would also work as `var x = (cp == 1 ? global.p2 : global.p1).show()` but that is a bit silly as there is no return value from the `show()` method.

**Reply by richardsg307** [reply]
I don't really understand what you are saying there as it's late (UK) and I've had a few whiskys, but if I add the `;` after the `print()` your code works!

**Comment by s-hadinger** [comment]
Fow what it's worth, I added `haspmota.page_show()` so you can trigger next/prev page.

https://github.com/arendst/Tasmota/pull/20333

```berry
haspmota.page_show('next')      # move to next page
haspmota.page_show('prev')      # move to previous page
haspmota.page_show('back')      # move to home page
haspmota.page_show('p3')        # move to page 3
```
